### PR TITLE
refactor(create-expo): swap `node-fetch` for built-in `fetch`

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This is a retry for PR #29391, but with a smaller scope (`create-expo`).

## Result

without `undici` or `node-fetch` | with `undici` | with `node-fetch`
--- | --- | ---
without-undici-or-node-fetch | with-undici | with-node-fetch

# How

- Replaced `node-fetch` with `undici` _(we probably can even drop undici too, but need to test that first)_
- Added `@octokit/types` for typed github responses

# Test Plan

See if CI passes for node 18, 20, and 22

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/create-expo/src/Examples.ts [hunk 1]
```diff
@@ -96,8 +96,15 @@ export async function downloadAndExtractExampleAsync(root: string, name: string)
...
+    debug(
+      `Failed to fetch the examples code, Github responded without content, received status "${response.status}"`
+    );
+    throw new Error('Failed to fetch the examples code from https://github.com/expo/examples');
+  }
+
   await pipeline(
-    response.body,
+    Readable.from(response.body),
     tar.extract(
```
Comment (byCedric) [inline packages/create-expo/src/Examples.ts:108]: This is a great example of converting web streams to node streams, when necessary.

**Comment by byCedric** [comment]
Found the culprit, we are hitting a fixed Terser issue: https://github.com/terser/terser/issues/1516

Specifically, `undici` contains a class called `FastIterableIterator` ([here](https://github.com/nodejs/undici/blob/2c9bc0b74f914a120a0497efad791b471767dc55/lib/web/fetch/util.js#L855)). This class [has a method](https://github.com/nodejs/undici/blob/2c9bc0b74f914a120a0497efad791b471767dc55/lib/web/fetch/util.js#L882C7-L882C75) that uses the following syntax:
```
if (typeof this !== 'object' || this === null || !(#target in this))
```

The current terser version transforms this into the following bad code:
```
// Bad, syntax error
if(typeof this!=="object"||this===null||!#_ in this)

// Should be:
if(typeof this!=="object"||this===null||!(#_ in this))
```

File: packages/create-expo/package.json
Comment (byCedric) [inline packages/create-expo/package.json:1]: We could add the `engines.node: >=18`, which warns users automatically when running it on deprecated Node versions.

**Review by Kudo — APPROVED** [review]

**Review by EvanBacon — APPROVED** [review]

File: packages/create-expo/package.json [hunk 2]
```diff
@@ -21,6 +21,9 @@
     "build",
     "template"
   ],
+  "engines": {
+    "node": ">=18"
```
Comment (byCedric) [inline packages/create-expo/package.json:25]:
```suggestion
    "node": ">=18.13.0"
```

This also includes the [`File`](https://nodejs.org/api/buffer.html#class-file) primitive in Node, which might come in handy when doing more advanced streaming.

File: packages/create-expo/CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 
 ### 🛠 Breaking changes
 
+- Removed `node-fetch` in favor of Node's built-in `fetch`. ([#29422](https://github.com/expo/expo/pull/29422) by [@byCedric](https://github.com/byCedric))
```
Comment (byCedric) [inline packages/create-expo/CHANGELOG.md:7]:
```suggestion
- Bump minimum required Node version to `18.13.0`. ([#29422](https://github.com/expo/expo/pull/29422) by [@byCedric](https://github.com/byCedric))
- Removed `node-fetch` in favor of Node's built-in `fetch`. ([#29422](https://github.com/expo/expo/pull/29422) by [@byCedric](https://github.com/byCedric))
```

**Pull request merged**